api/v1alpha1/healthcheck_types.go

```diff
@@ -0,0 +1,163 @@
+type HealthChecker struct {
+	// Type defines the type of health checker.
+	// +kubebuilder:validation:Enum=HTTP;GRPC;TCP
+	// +unionDiscriminator
+	Type HealthCheckerType `json:"type" yaml:"type"`
+	// HTTP defines the configuration of http health checker.
+	// It's required while the health checker type is HTTP.
+	// +optional
+	HTTP *HTTPHealthChecker `json:"http,omitempty" yaml:"http,omitempty"`
+	// GRPC defines the configuration of grpc health checker.
+	// It's required while the health checker type is GRPC.
+	// +optional
+	GRPC *GRPCHealthChecker `json:"grpc,omitempty" yaml:"grpc,omitempty"`
+	// TCP defines the configuration of tcp health checker.
+	// It's required while the health checker type is TCP.
+	// +optional
+	TCP *TCPHealthChecker `json:"tcp,omitempty" yaml:"tcp,omitempty"`
+}
+
+// HTTPHealthChecker defines the settings of http health check.
+type HTTPHealthChecker struct {
+	// Path defines the HTTP path that will be requested during health checking.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=1024
+	Path string `json:"path" yaml:"path"`
+	// Method defines the HTTP method used for health checking.
+	// Defaults to GET
+	// +optional
+	Method *string `json:"method,omitempty" yaml:"method,omitempty"`
+	// ExpectedStatuses defines a list of HTTP response statuses considered healthy.
+	// +optional
+	ExpectedStatuses []HTTPStatusRange `json:"expectedStatuses,omitempty" yaml:"expectedStatuses,omitempty"`
+	// ExpectedResponses defines a list of HTTP expected responses to match.
+	// +optional
+	ExpectedResponses []HealthCheckPayload `json:"expectedResponses,omitempty" yaml:"expectedResponses,omitempty"`
+}
+
+// GRPCHealthChecker defines the settings of grpc health check.
+type GRPCHealthChecker struct {
+	// ServiceName defines the value of parameter service in the gRPC health check request.
+	// +optional
+	ServiceName *string `json:"serviceName,omitempty" yaml:"omitempty"`
+	// Authority defines the value of :authority header in the gRPC health check request.
+	// +optional
+	Authority *string `json:"authority,omitempty" yaml:"authority,omitempty"`
+	// +optional
+	Metadata map[string]string `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+}
+
+// TCPHealthChecker defines the settings of tcp health check.
+type TCPHealthChecker struct {
+	// Send defines the request payload.
+	// +optional
+	Send *HealthCheckPayload `json:"send,omitempty" yaml:"send,omitempty"`
+	// Receive defines the expected response payloads.
+	// +optional
+	Receive []HealthCheckPayload `json:"receive,omitempty" yaml:"receive,omitempty"`
+}
+
+// HTTPStatusRange defines the start and end of the http status range using half-open interval semantics [start, end). Only statuses in the range [100, 600) are allowed.
+// +kubebuilder:validation:XValidation:rule="self.start <= self.end",message="start should be not greater than end"
+type HTTPStatusRange struct {
```

Comment:
imo prefer a list of values instead of range

Reply:
OK. I'll fix it later.

Reply:
this is an API change, changes can be backwards compatible, we should pick the most common approach

Reply:
done